File: lab01c.py
#!/usr/bin/env python3

def read_words(file_name):
    """
    Read words from a file.

    A word is any sequence of non-blank characters.
    Line breaks in the input are discarded.
    Stops when the character % is encountered.
    """
    with open(file_name) as file:
        for line in file:
            (line, eof, _) = line.partition('%')
            yield from line.split()
            if eof:  # found end-of-file character
                break


def print_words(file_name):
    """
    Read a sequence of words from a file and print them.

    One blank is printed between two words on the same line.
    A new line is started after any word that starts with r, s, or t.
    """
    line = []
    for word in read_words(file_name):
        line.append(word)
        if (word.startswith(('r', 's', 't'))):
            print(' '.join(line))
            line.clear()
    if line:
        print(' '.join(line))


print_words('lab03.txt')

File: src/bin/lab01c.rs
//! Programming Assignment 1: Part Two

use std::fs;

/// Read a sequence of words from a file and print them.
/// Line breaks in the input are discarded.
/// Any input after the % character is discarded.
/// A new line is started after any word that starts with r, s, or t.
/// One blank is printed between words on the same line.
fn main() {
    let file_name = "lab01c.txt";
    let mut content = String::new(); //initially empty
    if let Ok(file_content) = fs::read_to_string(file_name) {
        content = file_content;
    }
    if let Some(new_length) = content.find('%') {
        content.truncate(new_length);
    }
    let mut new_line = true;
    for word in content.split_whitespace() {
        if !new_line {
            print!(" ");
        }
        print!("{word}");
        new_line = word.starts_with(&['r', 's', 't']);
        if new_line {
            println!();
        }
    }
    if !new_line {
        println!();
    }
}
